applications/StructuralMechanicsApplication/custom_constitutive/hyper_elastic_isotropic_neo_hookean_3d.cpp

```diff
@@ -0,0 +1,441 @@
+
+    //Set the strain size
+    rFeatures.mStrainSize = GetStrainSize();
+
+    //Set the spacedimension
+    rFeatures.mSpaceDimension = WorkingSpaceDimension();
+}
+
+//************************************************************************************
+//************************************************************************************
+
+int HyperElasticIsotropicNeoHookean3D::Check(
+    const Properties& rmaterial_properties,
+    const GeometryType& rElementGeometry,
+    const ProcessInfo& rCurrentProcessInfo
+)
+{
+
+    if(YOUNG_MODULUS.Key() == 0 || rmaterial_properties[YOUNG_MODULUS] <= 0.0)
+    {
+        KRATOS_ERROR << "YOUNG_MODULUS has Key zero or invalid value " << std::endl;
+    }
+
+    const double& nu = rmaterial_properties[POISSON_RATIO];
+    const bool check = bool( (nu >0.499 && nu<0.501 ) || (nu < -0.999 && nu > -1.01 ) );
+
+    if(POISSON_RATIO.Key() == 0 || check==true)
+    {
+        KRATOS_ERROR << "POISSON_RATIO has Key zero invalid value " << std::endl;
+    }
+
+    if(DENSITY.Key() == 0 || rmaterial_properties[DENSITY] < 0.0)
+    {
+        KRATOS_ERROR << "DENSITY has Key zero or invalid value " << std::endl;
+    }
+
+    return 0;
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void HyperElasticIsotropicNeoHookean3D::CalculateConstitutiveMatrix(
+    Matrix& ConstitutiveMatrix,
+    const Matrix& InverseCTensor,
+    const double& DeterminantF,
+    const double& LameLambda,
+    const double& LameMu
+    )
+{
+    ConstitutiveMatrix.clear();
+    
+    const double log_j = std::log(DeterminantF);
+    
+    for(unsigned int i = 0; i < 6; i++)
+    {
+        for(unsigned int j = 0; j < 6; j++)
+        {
+            const unsigned int& i0 = this->msIndexVoigt3D6C[i][0];
+            const unsigned int& i1 = this->msIndexVoigt3D6C[i][1];
+            const unsigned int& j0 = this->msIndexVoigt3D6C[j][0];
+            const unsigned int& j1 = this->msIndexVoigt3D6C[j][1];
+            
+            ConstitutiveMatrix(i,j) = (LameLambda*InverseCTensor(i0,i1)*InverseCTensor(j0,j1)) + ((2.0 * LameMu-LameLambda * 2.0 * log_j) * 0.5 * (InverseCTensor(i0,j0) * InverseCTensor(i1,j1) + InverseCTensor(i0,j1) * InverseCTensor(i1,j0)));
+        }
+    }
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void HyperElasticIsotropicNeoHookean3D::CalculatePK2Stress(
+    const Matrix& CTensor,
+    Vector& rStressVector,
+    const double& DeterminantF,
+    const double& LameLambda,
+    const double& LameMu
+    )
+{
+    Matrix stress_matrix;
+    
+    const unsigned int dimension = WorkingSpaceDimension();
+    
+    stress_matrix  = LameLambda * std::log(DeterminantF) * CTensor + LameMu * ( IdentityMatrix(dimension, dimension) - CTensor );
```

Comment:
i don't think this formula is correct in large deformations.

you can not say that

stress = prod(stress_matrix, strain)

can u show me where you took the implementation from? because the log suggests me that you are using the one in logarithmic strains instrad of the one written in terms of cauchy green strains

Reply:
It is taken from Belyschko book, I will send you the screen shot later

Reply:
![imagen](https://user-images.githubusercontent.com/19991680/30251645-0e7ba858-9664-11e7-9551-987da28a7fe7.png)

Reply:
Good, but then you need to use 5.4.55 for the calculation of the stress and separately compute the elasticity tensor. When the strain is large you will have that for the exact tangent

D'S = elastic_tensor * Dstrain

But this will only be true for the increments, not for their total value

Reply:
The elastic tensor is: 

![imagen](https://user-images.githubusercontent.com/19991680/30288770-2de16bd0-972a-11e7-9dac-ce7b3e7ff339.png)

The piece of code corresponding to that is:

![imagen](https://user-images.githubusercontent.com/19991680/30328055-861c61d4-97ce-11e7-99ac-b99621c17cab.png)

Reply:
Do you mean something like in https://github.com/KratosMultiphysics/Kratos/blob/master/applications/structural_application/constitutive_laws/neo_hookean_3d.cpp ?

Reply:
your CalculateTangentMatrix/CalculateStress now looks a little better.

ensure to calculate the log just once since it is very expensive

Reply:
You can check the history, I just clean up, but the expression is still the same